File: app/src/main/java/com/example/nates/myfirstapp/MainActivity.java
package com.example.nates.myfirstapp;

import android.Manifest;
import android.app.Activity;
import android.content.pm.PackageManager;
import android.media.MediaPlayer;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.TextView;

import com.orbotix.ConvenienceRobot;
import com.orbotix.DualStackDiscoveryAgent;
import com.orbotix.common.DiscoveryException;
import com.orbotix.common.Robot;
import com.orbotix.common.RobotChangedStateListener;
import com.orbotix.le.RobotLE;

import java.util.HashMap;
import java.util.Map;


public class MainActivity extends Activity implements RobotChangedStateListener {

    private RobotActions mRobotActions;
    private RobotDances mRobotDances;
    private ConvenienceRobot mRobot;
    private DualStackDiscoveryAgent mDiscoveryAgent;
    private Map<String, Integer> clicks = new HashMap<>();
    private MediaPlayer mp;
    private final int CLICKSTOSTOP = 2;

    private static final int REQUEST_CODE_LOCATION_PERMISSION = 42;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        setupButtons();

        mDiscoveryAgent = new DualStackDiscoveryAgent();
        mDiscoveryAgent.addRobotStateListener(this);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        switch (requestCode) {
            case REQUEST_CODE_LOCATION_PERMISSION: {
                for (int i = 0; i < permissions.length; i++) {
                    if (grantResults[i] == PackageManager.PERMISSION_GRANTED) {
                        startDiscovery();
                        Log.d("Permissions", "Permission Granted: " + permissions[i]);
                    } else if (grantResults[i] == PackageManager.PERMISSION_DENIED) {
                        Log.d("Permissions", "Permission Denied: " + permissions[i]);
                    }
                }
            }
            break;
            default: {
                super.onRequestPermissionsResult(requestCode, permissions, grantResults);
            }
        }
    }


    @Override
    protected void onStart() {
        super.onStart();

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (checkSelfPermission(Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                return;
            }
        }
        startDiscovery();
    }

    private void startDiscovery() {
        final TextView textViewToChange = (TextView) findViewById(R.id.status_text);
        textViewToChange.setText("Connecting...");
        //If the DiscoveryAgent is not already looking for robots, start discovery.
        if (!mDiscoveryAgent.isDiscovering()) {
            try {
                mDiscoveryAgent.startDiscovery(getApplicationContext());
            } catch (DiscoveryException e) {
                Log.e("Sphero", "DiscoveryException: " + e.getMessage());
            }
        }
    }

    @Override
    protected void onStop() {
        //If the DiscoveryAgent is in discovery mode, stop it.
        if (mDiscoveryAgent.isDiscovering()) {
            mDiscoveryAgent.stopDiscovery();
        }

        //If a robot is connected to the device, disconnect it
        if (mRobot != null) {
            mRobot.disconnect();
            mRobot = null;
        }

        super.onStop();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mRobotActions.setRobotToDefaultState();
        mDiscoveryAgent.addRobotStateListener(null);
        if (mp.isPlaying()) {
            mp.stop();
            mp.release();
        }
    }

    @Override
    public void handleRobotChangedState(Robot robot, RobotChangedStateNotificationType type) {
        switch(type) {
            case Online: {
                final TextView textViewToChange = (TextView) findViewById(R.id.status_text);
                textViewToChange.setText("Connected!");

                //If robot uses Bluetooth LE, Developer Mode can be turned on.
                //This turns off DOS protection. This generally isn't required.
                if (robot instanceof RobotLE) {
                    ((RobotLE) robot).setDeveloperMode(true);
                }

                //Save the robot as a ConvenienceRobot for additional utility methods
                mRobot = new ConvenienceRobot(robot);
                mRobotActions = new RobotActions(mRobot);
                mRobotDances = new RobotDances(mRobot);

                Button runMacroButton = (Button) findViewById(R.id.run_macro);
                Button spinButton = (Button) findViewById(R.id.spin_button);
                runMacroButton.setEnabled(true);
                spinButton.setEnabled(true);

                break;
            }
            case Offline: {
                Button runMacroButton = (Button) findViewById(R.id.run_macro);
                Button spinButton = (Button) findViewById(R.id.spin_button);
                runMacroButton.setEnabled(false);
                spinButton.setEnabled(false);
            }
        }
    }

    private void setupButtons() {
        Button runMacroButton = (Button) findViewById(R.id.run_macro);
        runMacroButton.setOnClickListener( new OnClickListener() {
            @Override
            public void onClick(View v) {
                mRobotActions.runMacro();
            }
        });

        Button spinButton = (Button) findViewById(R.id.spin_button);
        spinButton.setOnClickListener( new OnClickListener() {
            @Override
            public void onClick(View v) {
                mRobotActions.spin();
            }
        });

        runMacroButton.setEnabled(false);
        spinButton.setEnabled(false);

        ImageButton playCheckup = (ImageButton) findViewById(R.id.play_checkup);
        playCheckup.setOnClickListener( new OnClickListener() {
            @Override
            public void onClick(View v) {
                triggerTimeForYourCheckup();
            }
        });

        ImageButton playDaniel = (ImageButton) findViewById(R.id.play_daniel);
        playDaniel.setOnClickListener( new OnClickListener() {
            @Override
            public void onClick(View v) {
                triggerDanielTiger();
            }
        });

        ImageButton playSesame = (ImageButton) findViewById(R.id.play_sesame);
        playSesame.setOnClickListener( new OnClickListener() {
            @Override
            public void onClick(View v) {
                triggerSesameStreet();
            }
        });

        ImageButton playElmo = (ImageButton) findViewById(R.id.play_elmo);
        playSesame.setOnClickListener( new OnClickListener() {
            @Override
            public void onClick(View v) {
                triggerElmosSong();
            }
        });

        ImageButton playSpider = (ImageButton) findViewById(R.id.play_spider);
        playSpider.setOnClickListener( new OnClickListener() {
            @Override
            public void onClick(View v) {
                triggerItsyBitsySpider();
            }
        });

        ImageButton playHead = (ImageButton) findViewById(R.id.play_head);
        playHead.setOnClickListener( new OnClickListener() {
            @Override
            public void onClick(View v) {
                triggerHeadShouldersKneesToes();
            }
        });
    }

    private void recordClick(String buttonName) {
        boolean keyFound = false;
        for (String key : clicks.keySet()) {
            if (key.equals(buttonName))
            {
                clicks.put(key, clicks.get(key) + 1);
                keyFound = true;
            }
            else
            {
                clicks.put(key, 0);
            }
        }

        if (!keyFound)
        {
            clicks.put(buttonName, 0);
        }
    }

    private void triggerSong(String song, int resid) {
        if (mp == null || !mp.isPlaying()) {
            mp = MediaPlayer.create(getApplicationContext(), resid);
            mp.start();
            if (mRobot != null && mRobot.isConnected()) {
                mRobotActions.setRobotToDefaultState();
                switch (song) {
                    case "doc":
                        mRobotDances.timeForYourCheckupDance();
                        break;
                    case "daniel":
                        mRobotDances.danielTigerDance();
                        break;
                    case "sesame":
                        mRobotDances.sesameStreetDance();
                        break;
                    case "elmo":
                        mRobotDances.elmosSongDance();
                        break;
                    case "spider":
                        mRobotDances.itsyBitsySpiderDance();
                        break;
                    case "head":
                        mRobotDances.headShouldersKneesToesDance();
                        break;
                }
            }
        } else if (clicks.containsKey(song) && clicks.get(song) >= CLICKSTOSTOP){
            mRobotActions.setRobotToDefaultState();
            mp.stop();
            clicks.put(song, 0);
        } else {
            recordClick(song);
        }
    }

    private void triggerDanielTiger() {
        triggerSong("daniel", R.raw.daniel_tiger_theme);
    }

    private void triggerTimeForYourCheckup() {
        triggerSong("doc", R.raw.time_for_your_checkup);
    }

    private void triggerSesameStreet() {
        triggerSong("sesame", R.raw.seasame_street_theme);
    }

    private void triggerElmosSong() {
        triggerSong("elmo", R.raw.elmos_song);
    }

    private void triggerItsyBitsySpider() {
        triggerSong("spider", R.raw.itsy_bitsy_spider);
    }

    private void triggerHeadShouldersKneesToes() {
        triggerSong("head", R.raw.head_shoulders_knees_toes);
    }

}

File: app/src/main/java/com/example/nates/myfirstapp/MainActivity.kt
package com.example.nates.myfirstapp

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.media.MediaPlayer
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import android.view.Gravity
import com.orbotix.ConvenienceRobot
import com.orbotix.macro.MacroObject
import java.util.*
import android.support.v7.app.AppCompatActivity
import android.widget.*
import com.orbotix.common.RobotChangedStateListener


class MainActivity : AppCompatActivity(), RobotServiceListener, BluetoothServiceListener {

    private var mBoundService: RobotProviderService? = null
    private var mBoundBluetoothService: BluetoothControllerService? = null
    private var mRobotActions = RobotActions()
    private var mRobotDances = RobotDances()
    private var mRobot: ConvenienceRobot? = null
    private val clicks = HashMap<Int, Int>()
    private var mp = MediaPlayer()
    private val clicksToStop = 2
    private var mIsBound = false
    private var mIsBluetoothBound = false

    private val mConnection = object : ServiceConnection {
        override fun onServiceConnected(className: ComponentName, service: IBinder) {
            Log.e("Activity","onServiceConnected")
            mBoundService = (service as RobotProviderService.RobotBinder).service
            mIsBound = true
            mBoundService?.addListener(this@MainActivity)
            if (mBoundService?.hasActiveRobot() == true) {
                handleRobotAlreadyConnected(mBoundService!!.getRobot())
            } else {
                var toast = Toast.makeText(this@MainActivity, "Discovering...",
                        Toast.LENGTH_LONG)
                toast.setGravity(Gravity.TOP, 0, 0)
                toast.show()
            }
        }

        override fun onServiceDisconnected(className: ComponentName) {
            Log.e("Activity","onServiceDisconnected")
            mBoundService = null
            mIsBound = false
            mBoundService?.removeListener(this@MainActivity)
        }
    }

    private val mBluetoothConnection = object : ServiceConnection {
        override fun onServiceConnected(className: ComponentName, service: IBinder) {
            Log.e("Activity","onServiceConnected")
            mBoundBluetoothService = (service as BluetoothControllerService.BluetoothBinder).service
            mIsBluetoothBound = true
            mBoundBluetoothService?.addListener(this@MainActivity)
            if (mBoundBluetoothService?.hasActiveBluetooth() == true) {
                handleBluetoothChange(1)
            }
        }

        override fun onServiceDisconnected(className: ComponentName) {
            Log.e("Activity","onServiceDisconnected")
            mBoundBluetoothService = null
            mIsBluetoothBound = false
            mBoundBluetoothService?.removeListener(this@MainActivity)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        Log.e("Activity", "onCreate")
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setupButtons()
    }

    override fun onStart() {
        Log.e("Activity", "onStart")
        val intent = Intent(this@MainActivity, BluetoothControllerService::class.java)
        bindService(intent, mBluetoothConnection, Context.BIND_AUTO_CREATE)
        super.onStart()
    }

    override fun onStop() {
        super.onStop()
        Log.e("Activity", "onStop")
        if (mIsBound) {
            unbindService(mConnection)
        }
        if (mIsBluetoothBound) {
            unbindService(mBluetoothConnection)
        }
    }

    override fun onDestroy() {
        Log.e("Activity", "onDestroy")
        super.onDestroy()
        if (mp.isPlaying) {
            mp.stop()
            mp.release()
        }
    }

    override fun handleBluetoothChange(type: Int) {
        val intent = Intent(this@MainActivity, RobotProviderService::class.java)
        bindService(intent, mConnection, Context.BIND_AUTO_CREATE)
    }

    fun handleRobotAlreadyConnected(robot: ConvenienceRobot) {
        mRobot = robot
        val macrosActivityButton = findViewById(R.id.robot_macros) as Button
        macrosActivityButton.isEnabled = true
    }

    override fun handleRobotChange(robot: ConvenienceRobot, type: RobotChangedStateListener.RobotChangedStateNotificationType) {
        when (type) {
            RobotChangedStateListener.RobotChangedStateNotificationType.Online -> {
                Log.e("Activity", "handleRobotConnected")
                var toast = Toast.makeText(this@MainActivity, "Connected!",
                        Toast.LENGTH_LONG)
                toast.setGravity(Gravity.TOP, 0, 0)
                toast.show()
                handleRobotAlreadyConnected(robot)
            }
            RobotChangedStateListener.RobotChangedStateNotificationType.Offline -> {
                Log.e("Activity", "handleRobotDisconnected")
                mRobot = null
                val macrosActivityButton = findViewById(R.id.robot_macros) as Button
                macrosActivityButton.isEnabled = false
            }
            RobotChangedStateListener.RobotChangedStateNotificationType.Connecting -> {
                Log.e("Activity", "handleRobotConnecting")
                var toast = Toast.makeText(this@MainActivity, "Connecting..",
                        Toast.LENGTH_LONG)
                toast.setGravity(Gravity.TOP, 0, 0)
                toast.show()
            }
        }    }

    private fun setupButtons() {
        val macrosActivityButton = findViewById(R.id.robot_macros) as Button
        macrosActivityButton.setOnClickListener {
            val intent = Intent(this, RobotMacrosActivity::class.java)
            startActivity(intent)
        }

        macrosActivityButton.isEnabled = false

        mapButton(R.id.play_checkup, mRobotDances::timeForYourCheckupDance, R.raw.time_for_your_checkup)
        mapButton(R.id.play_daniel, mRobotDances::danielTigerDance, R.raw.daniel_tiger_theme)
        mapButton(R.id.play_sesame, mRobotDances::sesameStreetDance, R.raw.seasame_street_theme)
        mapButton(R.id.play_elmo, mRobotDances::elmosSongDance, R.raw.elmos_song)
        mapButton(R.id.play_spider, mRobotDances::itsyBitsySpiderDance, R.raw.itsy_bitsy_spider)
        mapButton(R.id.play_head, mRobotDances::headShouldersKneesToesDance, R.raw.head_shoulders_knees_toes)
    }

    private fun recordClick(buttonId: Int) {
        var keyFound = false
        for (key in clicks.keys) {
            if (key == buttonId) {
                clicks.put(key, clicks[key]!! + 1)
                keyFound = true
            } else {
                clicks.put(key, 0)
            }
        }

        if (!keyFound) {
            clicks.put(buttonId, 0)
        }
    }

    private fun triggerSong(song: () -> MacroObject, resid: Int) {
        if (!mp.isPlaying) {
            mp = MediaPlayer.create(applicationContext, resid)
            mp.start()
            if (mRobot?.isConnected == true) {
                mRobotActions.setRobotToDefaultState(mRobot!!)
                val macro = song()
                macro.setRobot(mRobot!!.robot)
                macro.playMacro()
            }
        } else if (clicks.containsKey(resid) && clicks[resid]!! >= clicksToStop) {
            if (mRobot?.isConnected == true) mRobotActions.setRobotToDefaultState(mRobot!!)
            mp.stop()
            clicks.put(resid, 0)
        } else {
            recordClick(resid)
        }
    }

    private fun mapButton(button: Int, dance: () -> MacroObject, song: Int)
    {
        val playSong = findViewById(button) as ImageButton
        playSong.setOnClickListener { triggerSong(dance, song) }
    }
}